wikipedia_cleanup/random_forest.py

```diff
@@ -0,0 +1,141 @@
+                    self.related_properties_lookup = pickle.load(f)
+                return True
+            else:
+                print("No cache found, recalculating model.")
+        except EOFError:
+            print("Caching failed, recalculating model.")
+        return False
+
+    def fit(
+        self, train_data: pd.DataFrame, last_day: datetime, keys: List[str]
+    ) -> None:
+        if self.use_hash:
+            possible_cached_mapping = self._calculate_cache_name(train_data)
+            if self._load_cache(possible_cached_mapping):
+                return
+        ilocs = train_data.groupby(keys, sort=False).count()["value_valid_from"]
+        # Assumption: data is already sorted by keys so
+        # no further sorting needs to be done
+        start = 0
+        for iloc in tqdm(ilocs):
+            if iloc == 1:
+                continue
+            # maybe dont train if we are below a trainsize threshold
+            sample = train_data.iloc[start : start + iloc]
+            start += iloc
+            sample = sample.drop(columns=keys).iloc[:-1]
+            X = sample[self.get_relevant_attributes()].drop(
+                columns=["value_valid_from", "days_until_next_change"]
+            )
+            y = sample["days_until_next_change"]
+            reg = RandomForestClassifier(
+                random_state=0, n_estimators=10, max_features="auto"
+            )
+            reg.fit(X, y)
+            self.regressors[sample["key"].iloc[0]] = reg
+            self.last_preds[sample["key"].iloc[0]] = [pd.Timestamp("1999-12-31"), 0]
+
+        if self.use_hash:
+            possible_cached_mapping.parent.mkdir(exist_ok=True, parents=True)
+            with open(possible_cached_mapping, "wb") as f:
+                pickle.dump(self.regressors, f)
+
+    def _calculate_cache_name(self, data: pd.DataFrame) -> Path:
+        hash_string = "".join(str(x) for x in [data.shape, data.head(2), data.tail(2)])
+        hash_id = hashlib.md5(hash_string.encode("utf-8")).hexdigest()[:10]
+        possible_cached_mapping = self.hash_location / hash_id
+        return possible_cached_mapping
+
+    def predict_timeframe(
+        self,
+        data_key: np.ndarray,
+        additional_data: np.ndarray,
+        columns: List[str],
+        first_day_to_predict: date,
+        timeframe: int,
+    ) -> bool:
+        if len(data_key) == 0:
+            return False
+        key_column_idx = columns.index("key")
+        data_key_item = data_key[0, key_column_idx]
+        if data_key_item not in self.regressors:
+            # checks if model has been trained for the key
+            # (it didnt if there was no traindata)
+            return False
+
+        value_valid_from_column_idx = columns.index("value_valid_from")
+        sample = data_key[-1, ...]
+        sample_value_valid_from = sample[value_valid_from_column_idx]
+        if self.last_preds[data_key_item][0] != sample_value_valid_from:
```

Comment:
Some more info on how this caching works would be great.

Reply:
Added comment on initializations